src/spatial.rs

```diff
@@ -9,14 +9,32 @@ thread_local! {
     static NEXT_SHAPES_CACHE: RefCell<HashMap<Vec<usize>, Vec<Vec<usize>>>> = RefCell::new(HashMap::new());
 }
 
+fn apply_mapping(positions: &[Vec<usize>], mapping: &HashMap<Vec<usize>, usize>) -> Vec<usize> {
+    positions
+        .iter()
+        .map(|pos| *mapping.get(pos).expect("Position not found in new dimensions"))
+        .collect()
+}
+
 pub fn remap(old_dims: &[usize], new_dims: &[usize]) -> Vec<usize> {
     let old_positions = indices_in_order(old_dims);
     let mapping = location_to_index_mapping(new_dims);
 
-    old_positions
+    apply_mapping(&old_positions, &mapping)
+}
+
+pub fn remap_for_up(old_dims: &[usize], position: usize) -> Vec<usize> {
+    let padded_positions = pad(old_dims, position);
+    let insert_at = old_dims.len() - position;
+    let new_dims = old_dims[..insert_at]
```

Comment:
You don't need these insert at tricks for new dims. New dims match old dims but has an extra 2 added on either side. It doesn't matter which side as old dims is all 2s. 

Simplify this